Fragment does'n work in custom JSX Component

### What version of Hono are you using?

4.5.1

### What runtime/platform is your app running on?

Deno

### What steps can reproduce the bug?

- create custom JSX Component 
```ts
function MyComponent() {
  return <></>;
}
```
- use this component
```ts
  return <MyComponent></MyComponent>;
```
and the app crash.

### What is the expected behavior?

should render Fragment without import

### What do you see instead?

```console
2024-08-06T08:11:01.631Z [ERROR] [client] TypeError: Cannot read properties of null (reading 'isEscaped')
    at JSXFunctionNode.toStringToBuffer (https://jsr.io/@hono/hono/4.5.1/src/jsx/base.ts:251:64)
    at JSXFunctionNode.toString (https://jsr.io/@hono/hono/4.5.1/src/jsx/base.ts:143:12)
    at html (https://jsr.io/@hono/hono/4.5.1/src/helper/html/index.ts:35:29)
```

### Additional information

if you import `import { Fragment } from "hono/jsx/jsx-runtime";` it works, but IDE not report any info that something is wrong

--- a/runtime_tests/deno-jsx/jsx.test.tsx
+++ b/runtime_tests/deno-jsx/jsx.test.tsx
@@ -28,6 +28,12 @@ Deno.test('JSX: Fragment', () => {
   assertEquals(fragment.toString(), '<p>1</p><p>2</p>')
 })
 
+Deno.test('JSX: Empty Fragment', () => {
+  const Component = () => <></>
+  const html = <Component />
+  assertEquals(html.toString(), '')
+})
+
 Deno.test('JSX: Async Component', async () => {
   const Component = async ({ name }: { name: string }) =>
     new Promise<HtmlEscapedString>((resolve) => setTimeout(() => resolve(<span>{name}</span>), 10))

--- a/src/jsx/base.ts
+++ b/src/jsx/base.ts
@@ -247,7 +247,10 @@ class JSXFunctionNode extends JSXNode {
       children: children.length <= 1 ? children[0] : children,
     })
 
-    if (res instanceof Promise) {
+    if (typeof res === 'boolean' || res == null) {
+      // boolean or null or undefined
+      return
+    } else if (res instanceof Promise) {
       if (globalContexts.length === 0) {
         buffer.unshift('', res)
       } else {

--- a/src/jsx/base.ts
+++ b/src/jsx/base.ts
@@ -15,7 +15,7 @@ import { domRenderers } from './intrinsic-element/common'
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
 export type Props = Record<string, any>
 export type FC<P = Props> = {
-  (props: P): HtmlEscapedString | Promise<HtmlEscapedString>
+  (props: P): HtmlEscapedString | Promise<HtmlEscapedString> | null
   defaultProps?: Partial<P> | undefined
   displayName?: string | undefined
 }
@@ -374,11 +374,11 @@ export const memo = <T>(
   component: FC<T>,
   propsAreEqual: (prevProps: Readonly<T>, nextProps: Readonly<T>) => boolean = shallowEqual
 ): FC<T> => {
-  let computed: HtmlEscapedString | Promise<HtmlEscapedString> | undefined = undefined
+  let computed: ReturnType<FC<T>> = null
   let prevProps: T | undefined = undefined
-  return ((props: T & { children?: Child }): HtmlEscapedString | Promise<HtmlEscapedString> => {
+  return ((props) => {
     if (prevProps && !propsAreEqual(prevProps, props)) {
-      computed = undefined
+      computed = null
     }
     prevProps = props
     return (computed ||= component(props))

--- a/src/jsx/index.test.tsx
+++ b/src/jsx/index.test.tsx
@@ -383,6 +383,24 @@ describe('render to string', () => {
         '<html><head><title>Home page</title></head><body><h1>Hono</h1><p>Hono is great</p></body></html>'
       )
     })
+
+    describe('Booleans, Null, and Undefined Are Ignored', () => {
+      it.each([true, false, undefined, null])('%s', (item) => {
+        const Component: FC = (() => {
+          return item
+        }) as FC
+        const template = <Component />
+        expect(template.toString()).toBe('')
+      })
+
+      it('falsy value', () => {
+        const Component: FC = (() => {
+          return 0
+        }) as unknown as FC
+        const template = <Component />
+        expect(template.toString()).toBe('0')
+      })
+    })
   })
 
   describe('style attribute', () => {

